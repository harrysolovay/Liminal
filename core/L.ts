import "./intrinsics/mod.ts"

import type { PromiseOr } from "../util/mod.ts"
import type { Assertion, DescriptionArg, DescriptionParam, MetadataHandle } from "./Annotation.ts"

export * from "./intrinsics/mod.ts"
export * from "./utility/mod.ts"

export function _<K extends symbol, T = string>(
  key: K,
  serializer?: (value: T) => string,
): DescriptionParam<K, T> {
  return Object.assign(
    (value: T): DescriptionArg<K, T> => ({
      type: "DescriptionArg",
      key,
      value,
      serializer,
    }),
    {
      type: "DescriptionParam" as const,
      key,
    },
  )
}

export function assert<T, A extends unknown[]>(
  description: string | ((...args: A) => string),
  f?: (value: T, ...args: A) => PromiseOr<void>,
): (...args: A) => Assertion<T, A> {
  return (...args) => ({
    type: "Assertion",
    description,
    f,
    args,
  })
}

export function metadata<T = undefined>(key: symbol): MetadataHandle<T> {
  return Object.assign(
    (value: T) => ({
      type: "Metadata",
      key,
      value,
    }),
    { key },
  ) as never
}
