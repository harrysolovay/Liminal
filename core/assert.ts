import { assert as assert_, assertArrayIncludes, assertEquals } from "@std/assert"
import type { PartialType, Type } from "./Type.ts"
import { TypeVisitor } from "./TypeVisitor.ts"

export interface Diagnostic {
  type: PartialType
  description?: string
  path: string
  value: unknown
  exception?: unknown
}

export namespace Diagnostic {
  export function toString({ description, exception, path, value }: Diagnostic): string {
    return `Assertion${description ? ` "${description}"` : ""} failed ${
      exception
        ? exception instanceof Error ? `with Error "${exception.name}"` : "with exception"
        : ""
    } on value \`${JSON.stringify(value)}\` at \`root${path}\`: ${
      exception instanceof Error ? exception.message : JSON.stringify(exception)
    }`
  }
}

export async function assert(type: Type<any>, value: unknown): Promise<void> {
  const ctx = new AssertContext([], [], "root", value)
  visit(ctx, type)
  const diagnostics = [
    ...ctx.structuralDiagnostics,
    ...await Promise.all(ctx.annotationDiagnostics ?? []).then((v) => v.filter((e) => !!e)),
  ]
  if (diagnostics.length) {
    throw new AggregateError(diagnostics.map(({ exception }) => exception))
  }
}

export function matchUnionMember(members: Array<PartialType>, value: unknown): PartialType | void {
  const queue = [...members]
  while (queue.length) {
    const current = queue.pop()!
    const diagnostics: Array<Diagnostic> = []
    visit(new AssertContext(diagnostics, undefined, "", value), current)
    if (diagnostics.length) {
      continue
    }
    return current
  }
}

export class AssertContext {
  constructor(
    readonly structuralDiagnostics: Array<Diagnostic>,
    readonly annotationDiagnostics: undefined | Array<Promise<Diagnostic | void>>,
    readonly path: string,
    readonly value: unknown,
    readonly junction?: number | string,
  ) {}

  descend = (value: unknown, junction?: number | string): AssertContext =>
    new AssertContext(
      this.structuralDiagnostics,
      this.annotationDiagnostics,
      this.path,
      value,
      junction,
    )
}

const visit = TypeVisitor<AssertContext, void>({
  hook(next, ctx, type) {
    const path = ctx.junction ? ctx.path : `${ctx.path}["${ctx.junction}"]`
    try {
      next(ctx, type)
    } catch (exception: unknown) {
      ctx.structuralDiagnostics.push({
        type,
        path,
        exception,
        value: ctx.value,
      })
    }
    const { annotationDiagnostics } = ctx
    if (annotationDiagnostics) {
      type.annotations
        .filter((annotation) => typeof annotation === "object" && annotation?.type === "Assertion")
        .forEach(({ f, description }) => {
          if (f) {
            annotationDiagnostics.push((async () => {
              try {
                const passed = await f(ctx.value)
                if (!passed) {
                  return {
                    type,
                    path,
                    description,
                    value: ctx.value,
                  }
                }
              } catch (exception: unknown) {
                return {
                  type,
                  path,
                  exception,
                  description,
                  value: ctx.value,
                }
              }
            })())
          }
        })
    }
  },
  null({ value }) {
    assert_(value === null)
  },
  boolean({ value }) {
    assert_(typeof value === "boolean")
  },
  integer({ value }) {
    assert_(Number.isInteger(value))
  },
  number({ value }) {
    assert_(typeof value === "number")
  },
  string({ value }) {
    assert_(typeof value === "string")
  },
  const({ value }, _1, _2, cmpValue) {
    assertEquals(value, cmpValue)
  },
  enum({ value }, _1, ...values) {
    assert_(typeof value === "string")
    assertArrayIncludes(values, [value])
  },
  array(ctx, _1, element) {
    assert_(Array.isArray(ctx.value))
    ctx.value.forEach((value, i) => visit(ctx.descend(value, i), element))
  },
  object(ctx, _1, fields) {
    const keys = Object.keys(fields).toSorted()
    const { value } = ctx
    assert_(typeof value === "object" && value !== null)
    assertEquals(keys, Object.keys(value).toSorted())
    keys.forEach((key) => visit(ctx.descend(value[key as never], key), fields[key]!))
  },
  transform(ctx, _1, from) {
    visit(ctx, from)
  },
  union(ctx, _1, ...members) {
    const match = matchUnionMember(members, ctx.value)
    assert_(match)
    visit(ctx, match)
  },
  ref(ctx, _1, get) {
    visit(ctx, get())
  },
})
