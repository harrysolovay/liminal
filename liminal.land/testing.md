# Liminal Testing <Badge type="warning" text="beta" />
