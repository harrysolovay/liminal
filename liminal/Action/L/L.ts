import type { StandardSchemaV1 } from "@standard-schema/spec"
import type { Falsy } from "../../util/Falsy.js"
import type { LMembers } from "./LMembers.js"

export { struct as L } from "./intrinsics/struct.js"

export interface L<I = any, O = I> extends LMembers<I, O>, StandardSchemaV1<I, O> {
  (template: TemplateStringsArray, ...substitutions: Array<string>): this
  (...values: Array<Falsy | string>): this
}

// TODO: use these to minimize signature in cases where I/O is the same
export type I = typeof I
export declare const I: unique symbol
