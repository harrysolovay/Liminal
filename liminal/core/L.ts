import type { Assertion, DescriptionArg, DescriptionParam, Metadata } from "./Annotation.ts"

export * from "./intrinsics/mod.ts"
export * from "./utility/mod.ts"

export function _<K extends symbol, T = string>(
  key: K,
  serializer?: (value: T) => string,
): DescriptionParam<K, T> {
  return Object.assign(
    (value: T): DescriptionArg<K, T> => ({
      type: "DescriptionArg",
      key,
      value,
      serializer,
    }),
    {
      type: "DescriptionParam" as const,
      key,
    },
  )
}

export function assertion<T, A extends unknown[]>(
  description: string | ((...args: A) => string),
  f?: (value: T, ...args: A) => void | Promise<void>,
): (...args: A) => Assertion<T, A> {
  return (...args) => ({
    type: "Assertion",
    description,
    f,
    args,
  })
}

export function metadata(key: symbol, value: unknown): Metadata {
  return {
    type: "Metadata",
    key,
    value,
  }
}
