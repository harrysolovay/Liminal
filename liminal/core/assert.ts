import { AssertionContext } from "./AssertionContext.ts"
import type { AnyType } from "./Type.ts"

export async function assert(this: AnyType, value: unknown): Promise<void> {
  const ctx = new AssertionContext("root", [], [])
  ctx.visit(this, value)
  const diagnostics = [
    ...ctx.structuralDiagnostics,
    ...await Promise
      .all(ctx.annotationDiagnostics ?? []).then((v) => v.filter((e) => !!e)),
  ]
  if (diagnostics.length) {
    throw new AggregateError(diagnostics.map(({ exception }) => exception))
  }
}
