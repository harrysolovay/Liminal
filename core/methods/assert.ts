import { assert as assert_, assertArrayIncludes, assertEquals } from "@std/assert"
import type { Diagnostic } from "../Diagnostic.ts"
import type { AnyType } from "../Type.ts"
import { TypeVisitor } from "../TypeVisitor.ts"

export function assert<T>(this: AnyType<T>, value: unknown): asserts value is T {
  const diagnostics: Array<Diagnostic> = []
  visit(new AssertContext(diagnostics, "root", value), this)
  if (diagnostics.length) {
    throw new AggregateError(diagnostics.map(({ exception }) => exception))
  }
}

export class AssertContext {
  constructor(
    readonly diagnostics: Array<Diagnostic>,
    readonly path: string,
    readonly value: unknown,
    readonly junction?: number | string,
  ) {}

  descend = (value: unknown, junction?: number | string): AssertContext =>
    new AssertContext(this.diagnostics, this.path, value, junction)
}

const visit = TypeVisitor<AssertContext, void>({
  hook(next, ctx, type) {
    const path = ctx.junction ? ctx.path : `${ctx.path}["${ctx.junction}"]`
    try {
      next(ctx, type)
    } catch (exception: unknown) {
      ctx.diagnostics.push({
        type,
        path,
        exception,
        value: ctx.value,
      })
    }
  },
  null({ value }) {
    assert_(value === null)
  },
  boolean({ value }) {
    assert_(typeof value === "boolean")
  },
  integer({ value }) {
    assert_(Number.isInteger(value))
  },
  number({ value }) {
    assert_(typeof value === "number")
  },
  string({ value }) {
    assert_(typeof value === "string")
  },
  const({ value }, _1, _2, cmpValue) {
    assertEquals(value, cmpValue)
  },
  enum({ value }, _1, ...values) {
    assert_(typeof value === "string")
    assertArrayIncludes(values, [value])
  },
  array(ctx, _1, element) {
    assert_(Array.isArray(ctx.value))
    ctx.value.forEach((value, i) => visit(ctx.descend(value, i), element))
  },
  object(ctx, _1, fields) {
    const keys = Object.keys(fields).toSorted()
    const { value } = ctx
    assert_(typeof value === "object" && value !== null)
    assertEquals(keys, Object.keys(value).toSorted())
    keys.forEach((key) => visit(ctx.descend(value[key as never], key), fields[key]!))
  },
  transform(ctx, _1, from) {
    visit(ctx, from)
  },
  union(ctx, _1, ...members) {
    const match = matchUnionMember(members, ctx.value)
    assert_(match)
    visit(ctx, match)
  },
  f(ctx, _1, get) {
    visit(ctx, get())
  },
})

export function matchUnionMember(members: Array<AnyType>, value: unknown): AnyType | void {
  const queue = [...members]
  while (queue.length) {
    const current = queue.pop()!
    const diagnostics: Array<Diagnostic> = []
    visit(new AssertContext(diagnostics, "", value), current)
    if (diagnostics.length) {
      continue
    }
    return current
  }
}
