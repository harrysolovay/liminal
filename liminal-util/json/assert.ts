import { assert as assert_ } from "./assert.ts"
import type { Value } from "./Value.ts"

export function assert(value: unknown): asserts value is Value {
  if (value === null) {
    return
  } else if (Array.isArray(value)) {
    value.every(assert)
  } else if (typeof value === "object") {
    Object.values(value).every(assert)
  } else {
    assert_(typeof value === "string" || typeof value === "number" || typeof value === "boolean")
  }
}
