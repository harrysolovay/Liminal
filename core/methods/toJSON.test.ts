import { assert, assertEquals } from "@std/assert"
import { assertSnapshot } from "@std/testing/snapshot"
import type { AssertTrue, IsExact, IsNever } from "conditional-type-checks"
import { testIntrinsics } from "testing"
import * as L from "../L.ts"
import type { Type } from "../Type.ts"

testIntrinsics("toJSON", assertTypeSnapshot, {
  null: [
    [L.null],
  ],
  const: [
    [L.const(L.string, "Hello Liminal!")],
  ],
  boolean: [
    [L.boolean],
  ],
  number: [
    [L.number],
  ],
  integer: [
    [L.integer],
  ],
  string: [
    [L.string],
  ],
  array: [
    [L.array(L.string)],
  ],
  object: [
    [
      L.object({
        a: L.boolean,
        b: L.number,
        c: L.string,
      }),
    ],
  ],
  enum: [
    [L.enum("A", "B", "C")],
  ],
  union: [
    [L.union(L.null, L.string)],
  ],
  ref: [
    [L.ref(() => L.string)],
  ],
  transform: [
    [L.transform(L.string, (value) => ({ value }))],
  ],
})

async function assertTypeSnapshot(t: Deno.TestContext, type: Type<unknown>): Promise<void> {
  await assertSnapshot(t, type.toJSON())
  const withContext = type`One.`
  await assertSnapshot(t, withContext.toJSON())
  await assertSnapshot(t, withContext`Two.`.toJSON())
}

Deno.test("toJSON Descriptions", async (t) => {
  await t.step("Params", () => {
    const P_ = Symbol()
    const P = L.DescriptionParam(P_)
    const E = L.string`P: ${P}`
    type _0 = AssertTrue<IsExact<typeof E["D"], typeof P_>>
    const R = L.array(E)`Root.`
    type _1 = AssertTrue<IsExact<typeof R["D"], typeof P_>>
    const RR = R(P("V"))
    type _2 = AssertTrue<IsNever<typeof RR["D"]>>
    const json0 = RR.toJSON()
    assertEquals(json0.description, "Root.")
    assert("items" in json0)
    assertEquals(json0.items.description, "P: V")
    const RRR = RR`Another: ${P}.`
    type _3 = AssertTrue<IsExact<typeof RRR["D"], typeof P_>>
    const RRRR = RRR(P("X"))
    type _4 = AssertTrue<IsNever<typeof RRRR["D"]>>
    const json1 = RRRR.toJSON()
    assertEquals(json1.description, "Another: X. Root.")
    assert("items" in json1)
    assertEquals(json1.items.description, "P: V")
    const RRRRR = RRRR(P("Unused"))
    type _5 = AssertTrue<IsNever<typeof RRRRR["D"]>>
    const json2 = RRRRR.toJSON()
    assertEquals(json2.description, "Another: X. Root.")
    assert("items" in json2)
    assertEquals(json2.items.description, "P: V")
  })
})
