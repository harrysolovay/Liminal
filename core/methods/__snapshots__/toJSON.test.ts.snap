export const snapshot = {};

snapshot[`toJSON > null 1`] = `
{
  description: undefined,
  type: "null",
}
`;

snapshot[`toJSON > null 2`] = `
{
  description: "One.",
  type: "null",
}
`;

snapshot[`toJSON > null 3`] = `
{
  description: "One. Two.",
  type: "null",
}
`;

snapshot[`toJSON > const 1`] = `
{
  const: "Hello Liminal!",
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > const 2`] = `
{
  const: "Hello Liminal!",
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > const 3`] = `
{
  const: "Hello Liminal!",
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > boolean 1`] = `
{
  description: undefined,
  type: "boolean",
}
`;

snapshot[`toJSON > boolean 2`] = `
{
  description: "One.",
  type: "boolean",
}
`;

snapshot[`toJSON > boolean 3`] = `
{
  description: "One. Two.",
  type: "boolean",
}
`;

snapshot[`toJSON > number 1`] = `
{
  description: undefined,
  type: "number",
}
`;

snapshot[`toJSON > number 2`] = `
{
  description: "One.",
  type: "number",
}
`;

snapshot[`toJSON > number 3`] = `
{
  description: "One. Two.",
  type: "number",
}
`;

snapshot[`toJSON > integer 1`] = `
{
  description: undefined,
  type: "integer",
}
`;

snapshot[`toJSON > integer 2`] = `
{
  description: "One.",
  type: "integer",
}
`;

snapshot[`toJSON > integer 3`] = `
{
  description: "One. Two.",
  type: "integer",
}
`;

snapshot[`toJSON > string 1`] = `
{
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > string 2`] = `
{
  description: "One.",
  type: "string",
}
`;

snapshot[`toJSON > string 3`] = `
{
  description: "One. Two.",
  type: "string",
}
`;

snapshot[`toJSON > array 1`] = `
{
  description: undefined,
  items: {
    description: undefined,
    type: "string",
  },
  type: "array",
}
`;

snapshot[`toJSON > array 2`] = `
{
  description: "One.",
  items: {
    description: undefined,
    type: "string",
  },
  type: "array",
}
`;

snapshot[`toJSON > array 3`] = `
{
  description: "One. Two.",
  items: {
    description: undefined,
    type: "string",
  },
  type: "array",
}
`;

snapshot[`toJSON > object 1`] = `
{
  additionalProperties: false,
  description: undefined,
  properties: {
    a: {
      description: undefined,
      type: "boolean",
    },
    b: {
      description: undefined,
      type: "number",
    },
    c: {
      description: undefined,
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`toJSON > object 2`] = `
{
  additionalProperties: false,
  description: "One.",
  properties: {
    a: {
      description: undefined,
      type: "boolean",
    },
    b: {
      description: undefined,
      type: "number",
    },
    c: {
      description: undefined,
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`toJSON > object 3`] = `
{
  additionalProperties: false,
  description: "One. Two.",
  properties: {
    a: {
      description: undefined,
      type: "boolean",
    },
    b: {
      description: undefined,
      type: "number",
    },
    c: {
      description: undefined,
      type: "string",
    },
  },
  required: [
    "a",
    "b",
    "c",
  ],
  type: "object",
}
`;

snapshot[`toJSON > enum 1`] = `
{
  description: undefined,
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`toJSON > enum 2`] = `
{
  description: "One.",
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`toJSON > enum 3`] = `
{
  description: "One. Two.",
  enum: [
    "A",
    "B",
    "C",
  ],
  type: "string",
}
`;

snapshot[`toJSON > union 1`] = `
{
  anyOf: [
    {
      description: undefined,
      type: "null",
    },
    {
      description: undefined,
      type: "string",
    },
  ],
  description: undefined,
}
`;

snapshot[`toJSON > union 2`] = `
{
  anyOf: [
    {
      description: undefined,
      type: "null",
    },
    {
      description: undefined,
      type: "string",
    },
  ],
  description: "One.",
}
`;

snapshot[`toJSON > union 3`] = `
{
  anyOf: [
    {
      description: undefined,
      type: "null",
    },
    {
      description: undefined,
      type: "string",
    },
  ],
  description: "One. Two.",
}
`;

snapshot[`toJSON > ref 1`] = `
{
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > ref 2`] = `
{
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > ref 3`] = `
{
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > transform 1`] = `
{
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > transform 2`] = `
{
  description: undefined,
  type: "string",
}
`;

snapshot[`toJSON > transform 3`] = `
{
  description: undefined,
  type: "string",
}
`;
